nautobot/extras/tests/test_dynamicgroups.py

```diff
@@ -0,0 +1,360 @@
+        group = self.groups[0]
+        base_url = reverse("dcim:device_list")
+        params = group.map.urlencode(group.filter)
+        url = f"{base_url}?{params}"
+        self.assertEqual(group.get_group_members_url(), url)
+
+    def test_get_filter_fields(self):
+        """Test `DynamicGroup.get_filter_fields()`."""
+        # New instances should return {} `content_type` is set.
+        new_group = DynamicGroup(name="Unsaved Group", slug="unsaved-group")
+        new_filter_fields = new_group.get_filter_fields()
+        self.assertEqual(new_filter_fields, {})
+
+        # Existing groups should have actual fields.
+        group = self.groups[0]
+        filter_fields = group.get_filter_fields()
+        self.assertIsInstance(filter_fields, dict)
+        self.assertNotEqual(filter_fields, {})
+        self.assertIn("name", filter_fields)
+
+    def test_generate_filter_form(self):
+        """Test `DynamicGroup.get_filter_fields()`."""
+        group = self.groups[0]
+        filter_fields = group.get_filter_fields()
+        form_class = group.generate_filter_form()
+        form = form_class(group.filter)
+
+        # Form should validate just fine from the group's filter
+        self.assertTrue(form.is_valid())
+
+        # Form instance should have identical field set to filter fields.
+        self.assertEqual(
+            sorted(form.fields),
+            sorted(filter_fields),
+        )
+
+    def test_get_initial(self):
+        """Test `DynamicGroup.get_initial()`."""
+        group1 = self.groups[0]  # Filter has `site`
+        group2 = self.groups[1]  # Filter has `name`
+
+        # Test that a CharField (e.g. `name`) gets flattened. We use group2 for this.
+        initial = group2.get_initial()
+        expected = {"name": "bogus"}
+        self.assertEqual(initial, expected)
+
+        # Otherwise, it just passes through the filter.
+        self.assertEqual(group1.get_initial(), group1.filter)
+
+    def test_set_filter(self):
+        """Test `DynamicGroup.set_filter()`."""
+        group = self.groups[0]
+
+        # Input can come from a form's cleaned_data, such as our generated form. In this case, the
+        # filter we set from the form should be identical to what was there already.
+        old_filter = group.filter
+        form_class = group.generate_filter_form()
+        form_class.prefix = None  # We don't want the prefix here right now.
+        form = form_class(group.filter)
+        form.is_valid()
+        group.set_filter(form.cleaned_data)
+        self.assertEqual(group.filter, old_filter)
+
+        # Now we'll do it using a manually crafted dict.
+        new_filter = {"interfaces": True}
+        group.set_filter(new_filter)
+        self.assertEqual(group.filter, new_filter)
+
+        # And a bad input
+        bad_filter = {"site": -42}
+        with self.assertRaises(ValidationError):
+            group.set_filter(bad_filter)
+
+        # Cleanup
+        group.refresh_from_db()
```

Comment:
Should be unnecessary?

Reply:
It's because I used `setUpTestData()` vs. `setUp()` so the objects persist for the entire class. Comment expanded.